src/runtime/proc.go

```diff
@@ -353,6 +353,120 @@ func Gosched() {
+// fairness or starvation-prevention: once a goroutine Yields(), it may remain
+// parked until the scheduler next has idle capacity. This means Yield can block
+// for unbounded durations in the presence of sustained over-saturation; callers
+// are responsible for deciding where to Yield() to avoid priority inversions.
+//
+// Yield will never park if the calling goroutine is locked to an OS thread.
+func Yield() {
+	// Common/fast case: do nothing if npidle is non-zero meaning there there is
+	// an idle P so no reason to yield this one. Doing only this check here keeps
+	// Yield inlineable (~70 of 80 as of writing).
+	if sched.npidle.Load() == 0 {
+		maybeYield()
+	}
+}
+
+// maybeYield is called by Yield if npidle is zero, meaning there are no idle Ps
+// and thus there may be work to which the caller should yield. Such work could
+// be on this local runq of the caller's P, on the global runq, in the runq of
+// some other P, or even in the form of ready conns waiting to be noticed by a
+// netpoll which would then ready runnable goroutines.
+//
+// Keeping this function extremely cheap is essential: it must be cheap enough
+// that callers can call it in very tight loops, as very frequent calls ensure a
+// task wishing to yield when work is waiting will do so promptly. Checking the
+// runq of every P or calling netpoll are too expensive to do in every call, so
+// given intent is to bound how long work may wait, such checks only need to be
+// performed after some amount of time has elapsed (e.g. 0.25ms). To minimize
+// overhead when called at a higher frequency, this elapsed time is checked with
+// an exponential backoff.
+//
+// runqs are checked directly with non-atomic reads rather than runqempty: being
+// cheap is our top priority and a microsecond of staleness is fine as long as
+// the check does not get optimized out of a calling loop body (hence noinline).
+//
+//go:noinline
+func maybeYield() {
+	gp := getg()
+
+	// Don't park while locked to an OS thread.
+	if gp.lockedm != 0 {
+		return
+	}
+
+	if p := gp.m.p.ptr(); p.runqhead != p.runqtail || p.runnext != 0 {
+		if gp.yieldchecks == 1 {
+			yieldPark()
+			return
+		}
+		// To avoid thrashing between yields, set yieldchecks to 1: if we yield
+		// right back and see this sentinel we'll park instead to break the cycle.
+		gp.yieldchecks = 1
+		goyield()
+		return
+	}
+
+	// If the global runq is non-empty, park in the global yieldq right away.
+	if !sched.runq.empty() {
+		yieldPark()
+		return
+	}
+
+	const yieldCountBits, yieldCountMask = 11, (1 << 11) - 1
+	const yieldEpochShift = 18 - yieldCountBits
+	gp.yieldchecks++
+	if count := gp.yieldchecks & yieldCountMask; (count & (count + 1)) == 0 {
+		prev := gp.yieldchecks &^ yieldCountMask
+		now := uint32(nanotime()>>yieldEpochShift) &^ yieldCountMask
+		if now != prev || count == yieldCountMask {
+			gp.yieldchecks = now
+
+			for i := range allp {
```

Comment:
Let's add a comment that this is looking for a P with a non-empty runq. We know at this point that the current P's runq is empty and the global runq is empty. That means the system is mostly idle, so it's ok to do this semi-expensive check.

Reply:
The additional comments were helpful.

Do you think the exponential-backoff in nanotime() "epoch" checks is useful? Or did you measure it as beneficial? The alternative would be to check every N calls. When the counter becomes saturated we're checking every 512th call. To be clear, I don't have a stronger aversion to the exponential-backoff, just wondering how you landed here.

Do I have this math correct: the time epoch is being encoded in 21 bits with units ~0.25ms which allows encoding ~550s worth of values before wrapping around. It would be possible for `now == prev` inadvertently if `Yield` calls from a goroutine are happening very infrequently. That seems exceptionally unlikely.

Was there science behind the number of bits used for the counter and the size of the time epoch?

Reply:
Oops, you're right that this is 1 in 1k checks of nanotime() when the counter gets saturated.

> worse case we'll just wait for it to increment once more before we yield.
"it to increment" is referring to the time epoch, right?

> The exp backoff I think is how we ensure we gracefully calibrate to a caller call rate we don't control.
This explanation was good. I'd add it to the comment about the exponential backoff.